core/string/ustring.cpp

```diff
@@ -2372,66 +2372,94 @@ int64_t String::to_int() const {
 	return integer * sign;
 }
 
-int64_t String::to_int(const char *p_str, int p_len) {
-	int to = 0;
-	if (p_len >= 0) {
-		to = p_len;
-	} else {
-		while (p_str[to] != 0 && p_str[to] != '.') {
-			to++;
-		}
+template <class C>
+int64_t parse_integer(const C *p_str, int p_len, bool p_clamp) {
+	int str_index = 0;
+	C src_char = p_str[str_index];
+
+	// String is empty no conversion can be performed.
+	if (p_len == 0 || src_char == '\0') {
+		return 0;
 	}
 
-	int64_t integer = 0;
+	// Determine if the entire digit sequence should be read.
+
+	int end_pos = 0;
+	bool complete_seq = true;
+	if (p_len > 0) {
+		end_pos = p_len; // Parsing will stop at the provided position.
+		complete_seq = false;
+	}
+
+	// Check the sign at the beginning of the string.
+
 	int64_t sign = 1;
+	if (src_char == '-') {
+		sign = -1;
+		str_index++;
+	} else if (src_char == '+') {
+		str_index++;
+	}
+	src_char = p_str[str_index];
 
-	for (int i = 0; i < to; i++) {
-		char c = p_str[i];
-		if (is_digit(c)) {
-			bool overflow = (integer > INT64_MAX / 10) || (integer == INT64_MAX / 10 && ((sign == 1 && c > '7') || (sign == -1 && c > '8')));
-			ERR_FAIL_COND_V_MSG(overflow, sign == 1 ? INT64_MAX : INT64_MIN, "Cannot represent " + String(p_str).substr(0, to) + " as a 64-bit signed integer, since the value is " + (sign == 1 ? "too large." : "too small."));
-			integer *= 10;
-			integer += c - '0';
+	const int64_t max_div_10 = INT64_MAX / 10;
+	const int64_t min_div_10 = INT64_MIN / 10;
+	int64_t digit = 0;
+	int64_t integer = 0;
+	bool is_out_of_range = false;
 
-		} else if (c == '-' && integer == 0) {
-			sign = -sign;
-		} else if (c != ' ') {
+	// Parsing stops upon detecting overflow/underflow, reaching the specified end,
+	// or after reading the entire digit sequence.
+	while (is_digit(src_char)) {
+		digit = (src_char - '0') * sign;
+
+		if (str_index == end_pos && !complete_seq) {
 			break;
 		}
-	}
 
-	return integer * sign;
-}
+		// Detect integer overflow/underflow
```

Comment:
```suggestion
		// Detect integer overflow/underflow.
```
Per the [comment style](https://docs.godotengine.org/en/latest/contributing/development/code_style_guidelines.html#comment-style-guide).